### ISSUE TITLE
eOS: SFML windows are STILL being marked as unresponsive

### ISSUE THREAD

**Issue by barutzinkow** [SFML/SFML#1431]
## eOS: SFML windows are *still* being marked as unresponsive

In issue #1089, now closed, a strange bug was described to occur on the Gnome platform, namely a tendency for (some) SFML windows to get flagged as unresponsive by the OS right after opening. After SFML implementation changed back to Xlib, this seemed to fix the unresponsive-flag bug as well for sufficiently many people, so the issue was closed consequently. 

However, on my system (elementary OS Loki - based on Ubuntu 16.04.3 LTS and GTK+ 3.18.9) the very same bug still exists, even though I am using SFML 2.4.2 which should be all on Xlib's side again. The OS flags (most of) my SFML windows as unresponsive and shows a message about it, even though the application is fully reacting at all times. 

Since this bug does not *always* occur but only in apps of a certain complexity, I tried polling events as early as possible in the main loop or even before entering it. This workaround occasionally helped "fixing" the problem in very small applications, but in a little bigger ones that take some 300ms or so to start up, it does not seem to make a difference and I am confronted with the "window unresponsive" message every time.

This can become particularly annoying because the message box shown by the OS contains a button that immediately kills the application, even if pressed accidentally - i.e. by a key press that was intended for the application (which, after all, is fully working by the time the message pops up!) - thus any help on this is greatly appreciated.

**Comment by LaurentGomila** [comment]
Do you mean that it happens if the initialization phase (ie. the time between startup and first event polling) is too long?

**Comment by barutzinkow** [comment]
In a way, yes. If I create a new SFML application from scratch that does nothing except opening a window and run an events loop, the bug is gone. But now as I add more stuff to the app that has to be initialized before the main loop is entered the first time, like loading some images or sounds, there comes a point from where on the bug is there almost every time. I am not speaking here of a real lot - it is enough that the first event poll is delayed by just some milliseconds more for the OS to complain. This is also the reason why the "workaround" I described in the original post (polling and queuing or processing events *before* the main loop) doesn't provide a solution for real-world applications - given a certain complexity, the OS's unresponsiveness flagging will always be faster no matter how early I start polling.

---- Laurent Gomila schrieb ----

**Comment by MarioLiebisch** [comment]
What you describe is the expected behavior (at least on Windows). You should probably draw/handle some kind of loading screen that properly processes the window messages.

**Comment by eXpl0it3r** [comment]
Make sure you're *actually* using SFML 2.4.2. By that I mean, check that the library that have been linked are SFML 2.4.2 and not some other old ones. And by checking, I mean checking, not just assuming. 😉 

I guess the "issues" is that the window manager will keep marking the window unresponsive, even if the window started to work properly again.

Since this is only an issue with one distro so far, it seems like it should be solvable/fixed on the upstream side. Maybe try to figure out why this unresponsive message pops up and see if there's a config value that could be raised.

As a fix on your end, you might want to for example use a second thread to read the data from disk, while in the main thread you have a basic loop that handles events and after that load the resources from memory in the main thread. But make sure you synchronize everything correctly. Alternatively you could have a loop and then load resources step by step, instead of all at once.

**Comment by barutzinkow** [comment]
Thanks to everyone for their responses! I have been busy with other things in the last couple of days, I am sorry. But the issue still persists, so let me briefly describe which steps I have taken in the meantime:

- As eXpl0it3r suggested, I firstly double-checked that I *am* actually using SFML 2.4.2. In fact, everything else would have been a major surprise since I am coding on a new machine, and the 2.4.2 version of SFML is the only one ever installed on it. Nevertheless, I checked `/usr/lib` and `/usr/local/lib` and made sure there are no other library versions installed.
- Then, I moved the creation of the main window to the very first line of `main()` and polluted the init procedure with repeated calls to `sf::Window::clear()`, `pollEvent()`, and `display()` every second line to see if it would make a difference. It doesn't.
- I also tried creating a second "splash screen" window handling events as suggested by MarioLiebisch, but that did not prevent the bug either.
- What I *haven't* tried so far is delegating the initialization procedures into another thread. I will try this option today or tomorrow and report on whether it helped or not.

**Comment by barutzinkow** [comment]
After having been little successful with the multithreading approach, I visited the SFML homepage once again and discovered that there is now version 2.5 released. So I installed it and - surprise! All of a sudden the issue is gone... 

I still don't know what exactly was responsible for this weird behaviour, though. Since I had to recompile SFML from source (to take advantage of some C++17 features, I am also using brand-new g++8), I am not even sure it is the update to SFML that fixed it - maybe there was some compiler incompatibility going on? 

Anyway: Now that it is fixed, thanks again for your help!
